snuba/migrations/snuba_migrations/stacktraces/0001_stacktraces.py

```diff
@@ -0,0 +1,92 @@
+    # primary key
+    Column("organization_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("transaction_id", UUID()),
+    Column("received", DateTime()),
+
+    # profiling data
+    Column("stacktrace", String(Modifiers(codecs=["LZ4HC(9)"]))),
+    Column("symbols", String(Modifiers(codecs=["LZ4HC(9)"]))),
+
+    # filtering data
+    Column("android_api_level", UInt(32, Modifiers(nullable=True))),
+    Column("device_classification", String(Modifiers(low_cardinality=True))),
+    Column("device_locale", String(Modifiers(low_cardinality=True))),
+    Column("device_manufacturer", String(Modifiers(low_cardinality=True))),
+    Column("device_model", String(Modifiers(low_cardinality=True))),
+    Column("device_os_build_number", String(Modifiers(low_cardinality=True))),
+    Column("device_os_name", String(Modifiers(low_cardinality=True))),
+    Column("device_os_version", String(Modifiers(low_cardinality=True))),
+    Column("duration_ns", UInt(64)),
+    Column("environment", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("error_code", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("error_description", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("platform", String(Modifiers(low_cardinality=True))),
+    Column("trace_id", UUID()),
+    Column("transaction_name", String(Modifiers(low_cardinality=True))),
+    Column("version", NamedTuple((("name", String()), ("code", String())))),
+
+    # internal data
+    Column("retention_days", UInt(16)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.STACKTRACES,
+                table_name="stacktraces_local",
+                columns=columns,
+                engine=table_engines.MergeTree(
+                    storage_set=StorageSetKey.STACKTRACES,
+                    order_by="(organiation_id, project_id, transaction_id, received)",
+                    ttl="received + toIntervalDay(retention_days)",
+                    settings={"index_granularity": "8192"},
+                ),
+            )
+        ]
+
+    def backwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.STACKTRACES, table_name="stacktraces_local",
+            )
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.STACKTRACES,
+                table_name="stacktraces_dist",
+                columns=columns,
+                engine=table_engines.Distributed(
+                    local_table_name="stacktraces_local",
+                    sharding_key="cityHash64(project_id, transaction_name)",
```

Comment:
I think it may be good to pick a sharding key that distributes the data a bit more evenly between the shards (Not sure what transaction_name represents exactly but it does seem to be a a low cardinality thing), e.g. `cityHash64(transaction_id)` or something like that. @fpacifici @evanh Any opinions here?

Reply:
Clickhouse is quite efficient at partitioning evenly up at least 10 shards (likely more). If you partition evenly as Lyn suggested you get a few perks for free:
- even distribution when you add shards
- no hot shards, queries on a massive project will be distributed across shards. if you put  a project and a transaction_name in the same shard, one host alone will have to perform the entire scan.

Now this will depend a lot on your query pattern. 
- if you query by project_id,transaction id, and shard by project_id, transaction_name, you do not take advantage of shard locality at all. Clickhouse needs to fan out the query to all shards anyway
- if you query by project_id and any field other than transaction_name you have the same problem. Clickhouse cannot narrow down one shard so it runs the query on all of them.

So it seems that you do not gain a lot by sharding by anything that is not spreading data evenly across shards. 
If you where sharding by project_id and transaciton_id then yes you would be able to narrow down the query to one shard knowing in advance where to find your data. Though you will not gain anything in the queries that are likely to be heavier in terms of load as they have to scan more data: aggregation by any other field.

Reply:
I guess my idea was to group data having the same transaction name as we'll aggregate by that a lot but I didn't really give enough thought to it. It sounds like a bad idea.

Given what you said, using the `transaction_id` will provide the most uniform distribution. Let's use that then.

Reply:
please keep the transaciton_id, not rand. We may be able to query only one shard if we know which transaction_id we look for and we partition by transaction_id.